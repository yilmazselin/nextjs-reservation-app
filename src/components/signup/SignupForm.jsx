import { useTranslation } from "next-i18next"
import { Formik } from "formik"
import { Input, SelectBox } from "../Form"
import Button from "../Buttons/Button"
import Loader from "../common/Loader"
import { SignupSchema, initialValues, sleep } from "./SignupFormHelper"
import { localeOptions } from "../../utils/constant"

function SignupForm({ onSubmit }) {
  const { t } = useTranslation("auth")

  const handleSubmit = async (values) => {
    await sleep(500)
    onSubmit(values)
  }

  return (
    <Formik
      initialValues={initialValues}
      onSubmit={handleSubmit}
      validationSchema={SignupSchema}
    >
      {({
        handleChange,
        values,
        handleSubmit,
        errors,
        touched,
        isSubmitting,
      }) => (
        <form autoComplete="off">
          <Input
            error={touched.email ? errors.email : null}
            id="email"
            label={t("common:labels.email")}
            marginBottom={40}
            name="email"
            onChange={handleChange}
            type="email"
            value={values.email}
          />
          <Input
            error={touched.password ? errors.password : null}
            id="password"
            label={t("common:labels.password")}
            marginBottom={45}
            name="password"
            onChange={handleChange}
            type="password"
            value={values.password}
          />
          <SelectBox
            defaultValue={values.locale}
            id="locale"
            label={t("common:labels.locale")}
            marginBottom={35}
            name="locale"
            onChange={handleChange}
            options={localeOptions}
          />
          <Button
            color="primary"
            disabled={
              isSubmitting ? isSubmitting : !values.password && !values.email
            }
            onClick={handleSubmit}
            role="submit-button"
            type="submit"
          >
            {!isSubmitting ? t("auth:signup-button") : <Loader />}
          </Button>
        </form>
      )}
    </Formik>
  )
}
export default SignupForm
