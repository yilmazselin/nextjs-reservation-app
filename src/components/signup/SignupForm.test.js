import "@testing-library/jest-dom"
import {
  act,
  fireEvent,
  render,
  screen,
  waitFor
} from "@testing-library/react"
import userEvent from "@testing-library/user-event"
import "jest-styled-components"
import { useTranslation } from "react-i18next"
import SignupForm from "./SignupForm"

jest.mock("react-i18next", () => ({
  useTranslation: jest.fn()
}))

const tSpy = jest.fn((str) => str)

useTranslation.mockReturnValue({
  t: tSpy,
  i18n: {
    changeLanguage: () => new Promise(() => {})
  }
})

describe("<SignupForm />", () => {
  it("should render the form fields correctly", () => {
    const { container } = render(<SignupForm onSubmit={false} />)

    expect(container).toMatchSnapshot()
    expect(screen.getByLabelText(/common:labels.email/i)).toBeInTheDocument()
    expect(screen.getByLabelText(/common:labels.password/i)).toBeInTheDocument()
    expect(screen.getByLabelText(/common:labels.locale/i)).toBeInTheDocument()
    const button = screen.getByRole("submit-button")

    expect(button).toBeInTheDocument()
    expect(button).toBeDisabled()
  })

  it("should submit the form data correctly", async () => {
    const handleSubmit = jest.fn()

    render(<SignupForm onSubmit={handleSubmit} />)
    const user = userEvent.setup()

    const email = screen.getByLabelText(/common:labels.email/i)
    const password = screen.getByLabelText(/common:labels.password/i)
    const locale = screen.getByLabelText(/common:labels.locale/i)
    const submitButton = screen.getByRole("submit-button")

    await user.type(email, "test@gmail.com")
    await user.type(password, "123456")
    act(() => {
      fireEvent.change(locale, {
        target: {
          value: "en"
        }
      })
    })
    await user.click(submitButton)
    expect(submitButton).toBeDisabled()

    await waitFor(() => expect(handleSubmit).toHaveBeenCalledWith({
      email: "test@gmail.com",
      password: "123456",
      locale: "en"
    }))
  })
})
