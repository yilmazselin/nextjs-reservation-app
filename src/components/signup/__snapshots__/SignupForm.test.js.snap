// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SignupForm /> should render the form fields correctly 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  position: relative;
  margin-bottom: 40px;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  position: relative;
  margin-bottom: 45px;
}

.c3 {
  color: #64738C;
  font-size: 16px;
  font-family: "NunitoSans-500","sans-serif";
  -webkit-order: 1;
  -ms-flex-order: 1;
  order: 1;
  pointer-events: none;
  -webkit-transform-origin: left top;
  -ms-transform-origin: left top;
  transform-origin: left top;
  -webkit-transform: scale(1) translate3d(0,22px,0);
  -ms-transform: scale(1) translate3d(0,22px,0);
  transform: scale(1) translate3d(0,22px,0);
  -webkit-transition: 200ms ease all;
  transition: 200ms ease all;
  position: relative;
}

.c1 {
  position: relative;
  border-radius: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  color: #121212;
  font-size: 16px;
  font-family: "NunitoSans-600","sans-serif";
  line-height: 24px;
  padding-bottom: 8px;
  outline: 0;
  border: 0;
  border-bottom: 1px solid;
  border-bottom-color: #BBC3CF;
  -webkit-flex: 1 1 auto;
  -ms-flex: 1 1 auto;
  flex: 1 1 auto;
  -webkit-order: 2;
  -ms-flex-order: 2;
  order: 2;
  -webkit-transition: all 0.2s ease;
  transition: all 0.2s ease;
}

.c1:focus ~ .c2,
.c1:not(:placeholder-shown) ~ .c2 {
  color: #77869e;
  -webkit-transform: scale(0.8) translate3d(0,-8px,0);
  -ms-transform: scale(0.8) translate3d(0,-8px,0);
  transform: scale(0.8) translate3d(0,-8px,0);
}

.c1:not(:placeholder-shown) {
  border-bottom-color: #64738C;
}

.c1:focus {
  border-bottom-color: #0DAFC0;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  position: relative;
  margin-bottom: 35px;
  overflow: hidden;
}

.c6 {
  color: #64738C;
  font-size: 16px;
  font-family: "NunitoSans-500","sans-serif";
  margin-bottom: 10px;
  font-size: 13px;
}

.c7 {
  border: none;
  outline: none;
  border-bottom: 1px solid #64738C;
  padding-bottom: 8px;
  margin-left: -5px;
  width: calc(100% + 5px);
  color: #121212;
  font-size: 16px;
  font-family: "NunitoSans-600","sans-serif";
}

.c8 {
  width: 100%;
  height: 56px;
  border: none;
  border-radius: 12px;
  outline: none;
  -webkit-transition: all 0.2s;
  transition: all 0.2s;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  color: #fff;
  font-size: 14px;
  font-family: "NunitoSans-700","sans-serif";
  color: #fff;
  background-color: #E82223;
  cursor: default;
  background-color: #BBC3CF;
  color: #fff;
}

<div>
  <form
    autocomplete="off"
  >
    <div
      class="c0"
    >
      <input
        class="c1"
        id="email"
        name="email"
        placeholder=" "
        type="email"
        value=""
      />
      <label
        class="c2 c3"
        for="email"
      >
        common:labels.email
      </label>
    </div>
    <div
      class="c4"
    >
      <input
        class="c1"
        id="password"
        name="password"
        placeholder=" "
        type="password"
        value=""
      />
      <label
        class="c2 c3"
        for="password"
      >
        common:labels.password
      </label>
    </div>
    <div
      class="c5"
    >
      <label
        class="c6"
        for="locale"
      >
        common:labels.locale
      </label>
      <select
        class="c7"
        data-testid="select"
        id="locale"
        name="locale"
      >
        <option
          data-testid="select-option"
          selected=""
          value="tr"
        >
          common:languages.tr
        </option>
        <option
          data-testid="select-option"
          value="en"
        >
          common:languages.en
        </option>
      </select>
    </div>
    <button
      class="c8"
      color="primary"
      disabled=""
      role="submit-button"
      type="submit"
    >
      auth:signup-button
    </button>
  </form>
</div>
`;
